### PR TITLE
Initial draft DPoP with Angular

### PR DESCRIPTION
**Pull request by alisaduncan**
This blog post has the following features:

- [x] A GitHub Repository with a polished README
- [x] A GitHub Repository under the github.com/oktadev account
- [x] A title that's approved by Dev Advocacy
- [x] A URL approved by Dev Advocacy
- [x] The content has been run through Grammarly (https://www.grammarly.com/)
- [x] Rendered locally and confirmed that no Markdown typos exist
- [x] Images are compressed appropriately
- [x] Social image previews well on Twitter and LinkedIn
- [x] Tech review request (developer advocate or domain expert)
- [x] Review request for editorial/grammar/clarity (developer advocate or Colton)

### CODE REVIEW

File: _source/_posts/2024-08-09-angular-dpop-jwt.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+tags: []
+tweets:
+- ""
+- ""
+- ""
+image:
+type: conversion
+---
+
+We use OAuth 2.0 access tokens to request data and to perform actions within our software systems. A resource server first checks the validity of the access token before acting upon the HTTP request. What happens if the requesting party is a malicious sort who stole your token? Would the resource server honor the HTTP request? If you use a bearer token, the answer is "Yes."
```
Comment (coltonbeach) [inline _source/_posts/2024-08-09-angular-dpop-jwt.md:17]:
```suggestion
We use OAuth 2.0 access tokens to request data and to perform actions within our software systems. A resource server first checks the validity of the access token before acting upon the HTTP request. What happens if the requesting party is a malicious sort who stole your token? Would the resource server honor the HTTP request? If you use a bearer token, the answer is "yes." 
```

File: _source/_posts/2024-08-09-angular-dpop-jwt.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  * Terminal window (if you aren't using an IDE with a built-in terminal)
+  * [Git](https://git-scm.com/) and an optional [GitHub account](https://github.com/) if you want to track your changes using a source control manager
+  
+> **Note**
+>
+> This code project is best for developers familiar with web development basics.
+>
+> Are you following along the step-by-step code instructions in Angular? If you are an Angular newbie, start by building [your first Angular app](https://angular.io/tutorial/first-app) using the tutorial created by the Angular team.
+
+**JAD TODO - Let's start with an application using OpenID Connect and OAuth 2.0** We're following best-practices, that is for this Angular app (a SPA) we're using authorization code flow with PKCE. Create the Okta developer account and get app going. Sign in, following network calls and URL redirection requests. Using Org AS, and making a call to users API.
```
Comment (coltonbeach) [inline _source/_posts/2024-08-09-angular-dpop-jwt.md:50]:
```suggestion
**JAD TODO - Let's start with an application using OpenID Connect and OAuth 2.0** We're following best-practices, that is for this Angular app (a SPA) we're using authorization code flow with PKCE. Create the Okta developer account and get app going. Sign in following network calls and URL redirection requests. Use Org AS, and make a call to users API.
```

File: _source/_posts/2024-08-09-angular-dpop-jwt.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+Now grab the access token and make the same call from a HTTP client. Uh-oh!
+
+## Binding OAuth 2.0 access tokens
+
+If we go back to the hotel keycard example, what we want is a hotel keycard only you can use. Something that links you are the rightful user of the hotel keycard.
```
Comment (coltonbeach) [inline _source/_posts/2024-08-09-angular-dpop-jwt.md:56]:
```suggestion
If we go back to the hotel keycard example, what we want is a hotel keycard only you can use; something that links you are the rightful user of the hotel keycard. 
```

File: _source/_posts/2024-08-09-angular-dpop-jwt.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+OAuth 2.0 defines a few methods to bind access tokens such as:
+  * [Client Secret](https://oauth.net/2/client-authentication/) where the token request includes a secret value only known to the requester and the authorization server
+  * [Mutual TLS Client Authentication and Certificate-Bound Access Tokens (mTLS)](https://oauth.net/2/mtls/) where the access tokens are bound to the web server's certificate
+  * [Private Key JSON Web Token (JWT)](https://oauth.net/private-key-jwt/) where the requesting machine-to-machine service application uses asymmetric encryption to sign a JWT
+
+The methods to bind access tokens apply only to confidential clients, those that can maintain secrets. Which means the methods don't apply to public clients. Public clients are apps that run authentication code within the user's hardware, such as in Single-Page Applications (SPA) and mobile clients. Publics clients are widely used with avenues for token security exploits without careful protection. Is there an alternative?
```
Comment (coltonbeach) [inline _source/_posts/2024-08-09-angular-dpop-jwt.md:65]:
```suggestion
The methods to bind access tokens apply only to confidential clients, (those that can maintain secrets) which means the methods don't apply to public clients. Public clients are apps that run authentication code within the user's hardware, such as in Single-Page Applications (SPA) and mobile clients. Public clients are widely used with avenues for token security exploits without careful protection. Is there an alternative?
```

File: _source/_posts/2024-08-09-angular-dpop-jwt.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+> **Note**
+> 
+> While we'll focus on SPAs in this post, the DPoP spec applies to any OAuth client and elevates authorization security by binding the token to the requestor. Some industries, such as finance, now recommend or require DPoP to protect API calls, regardless if the calling party is a SPA, mobile, or a machine-to-machine service application.
+
+
+**JAD TODO - What happens if the user tries authenticating without DPoP** 
+
+
+When using DPoP, the client uses asymmetric encryption to create the proof. The DPoP proof is a JWT that includes the URI and the HTTP method of the request along with the public key. (The proof binds the HTTP request to alleviate overpowered malicious parties who steal tokens.)???
```
Comment (coltonbeach) [inline _source/_posts/2024-08-09-angular-dpop-jwt.md:79]:
```suggestion
When using DPoP, the client uses asymmetric encryption to create the proof. The DPoP proof is a JWT that includes the URI and the HTTP method of the request along with the public key. (The proof binds the HTTP request to alleviate overpowered malicious parties who steal tokens.)
```

File: _source/_posts/2024-08-09-angular-dpop-jwt.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+The client includes the proof when requesting tokens from the authorization server. This allows the authorization server to bind the access tokens to the requesting client AND the URI plus HTTP method of the call. When the client requests resources, they send the access token and the proof they are the legitimate holder of the token. **JAD TODO - CLEAN UP**
+
+Now that token has possession proof, the tokens are no longer bearer tokens. Only the requesting client can use the token. Clients request resources by sending the access token along with a proof demonstrating they are the legitimate holders of the access token to resource servers using a new scheme `DPoP`. Let's walk through the process.
```
Comment (coltonbeach) [inline _source/_posts/2024-08-09-angular-dpop-jwt.md:83]:
```suggestion
Now that token has possession proof, the tokens are no longer bearer tokens. Only the requesting client can use the token. Clients request resources by sending the access token along with a proof demonstrating they are the legitimate holders of the access token to resource servers using a new scheme `DPoP`. Let's walk through the process:
```

**Review by coltonbeach — CHANGES_REQUESTED** [review]

File: _source/_posts/2024-09-10-angular-dpop-jwt.md [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+tags: [angular, react, vue, spa, authorization]
+image:
+type: conversion
+github: https://github.com/oktadev/okta-angular-dpop-example
+---
+
+In OAuth, a valid access token grants the caller access to resources and the ability to perform actions on the resources. This means the access token is powerful and dangerous if it falls into malicious hands. The traditional bearer token scheme means the token grants anyone who possesses it access. A new OAuth 2.0 extension specification, [Demonstrating Proof of Possession (DPoP)](https://www.rfc-editor.org/rfc/rfc9449), defines a standard way that binds the access token to the OAuth client sending the request, elevating access token security. 
+
+The high-level overview of DPoP uses public/private keys to create a signed DPoP proof that the authorization and resource server use to confirm the authenticity of the request and requesting client. This way, the token is sender-constrained, and a token thief can less use a compromised access token. Learn more about the problems DPoP solves and how it works by reading:
```
Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
The high-level overview of DPoP uses public/private keys to create a signed DPoP proof that the authorization and resource server use to confirm the authenticity of the request and requesting client. This way, the token is sender-constrained, and a token thief is less likely to use a compromised access token. Learn more about the problems DPoP solves and how it works by reading:
```

File: _source/_posts/2024-09-10-angular-dpop-jwt.md [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+
+{% excerpt /blog/2024/09/05/dpop-oauth %}
+
+The primary use case for DPoP is for public clients, the applications where authentication code runs within the end user's browser, such as Single-Page Applications (SPA) and mobile apps. Due to their architecture, public clients inherently have higher risk and less security in authentication and authorization. Public clients can't leverage a client secret used by application types that can communicate to the authorization server through a "back-channel," a network connection opaque to users, network sniffing attackers, and nosy developers. Without proper protection, a SPA may store tokens accessible to the end-user and injection-related attacks. DPoP adds an extra protection layer that makes tokens less usable if stolen.
```
Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
The primary use case for DPoP is for public clients, (the applications where authentication code runs within the end user's browser, such as Single-Page Applications (SPA) and mobile apps). Due to their architecture, public clients inherently have higher risk and less security in authentication and authorization. Public clients can't leverage a client secret used by application types that can communicate to the authorization server through a "back-channel," a network connection opaque to users, network sniffing attackers, and nosy developers. Without proper protection, a SPA may store tokens accessible to the end-user and injection-related attacks. DPoP adds an extra protection layer that makes tokens less usable if stolen.
```

File: _source/_posts/2024-09-10-angular-dpop-jwt.md [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+  * [Node.js](https://nodejs.org/en) v18 or greater
+  * A web browser with good debugging capabilities, such as Chrome
+  * Your favorite IDE. Still looking? I like VS Code and WebStorm because they have integrated terminal windows.
+  * Terminal window (if you aren't using an IDE with a built-in terminal)
+  * [Git](https://git-scm.com/) and an optional [GitHub account](https://github.com/) if you want to track your changes using a source control manager
+  * An HTTP client that shows the HTTP requests and responses, such as the [Http Client VS Code extension](https://marketplace.visualstudio.com/items?itemName=Aaron00101010.http-client) or [curl](https://curl.se/)
+
+## Get the starting Angular, React, or Vue project
+
+You'll use a starter project. The instructions are for the Angular sample project. If you are following along in [React](https://github.com/okta-samples/okta-react-sample) or [Vue](https://github.com/okta-samples/okta-vue-sample), replace the GitHub repo location with the URL for the sample you're using.
```
Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
You'll use a starter project. These instructions are for the Angular sample project. If you are following along in [React](https://github.com/okta-samples/okta-react-sample) or [Vue](https://github.com/okta-samples/okta-vue-sample), replace the GitHub repo location with the URL for the sample you're using. 
```

File: _source/_posts/2024-09-10-angular-dpop-jwt.md [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+
+The request includes the `Authorization` header containing the token scheme and access token. You see the format `Bearer <access_token>`.
+
+{% img blog/angular-dpop-jwt/bearer-network-users-api-request-headers.jpg alt:"Users API request headers showing the Authorization header." width:"800" %}{: .center-image }
+
+The entity holding the token can legitimately request resources. Let's try using the token in another client and impersonating the actions an attacker can take if they manage to capture it. 
+
+> **Note**
+>
+> Access tokens expire quickly. If too much time passes in these next steps, you may get a `401 Unauthorized`. If you do, repeat the steps with a more recent access token by navigating between the profile and user routes to trigger a call to the API. It prompts the OIDC client, the Okta Auth JS SDK, to update expired tokens.
```
Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
> Access tokens expire quickly. If too much time passes in these next steps, you may get a `401 Unauthorized`. If you do, repeat the steps with a more recent access token by navigating between the profile and user routes to trigger a call to the API. It prompts the OIDC client (the Okta Auth JS SDK) to update expired tokens.
```

File: _source/_posts/2024-09-10-angular-dpop-jwt.md [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+Authorization: Bearer {yourAccessToken}
+```
+
+If you use curl, add the verbose flag to see the request and response headers:
+
+```curl
+curl -v --header "Authorization: Bearer {yourAccessToken}" https://{yourOktaDomain}/api/v1/users
+```
+
+The call succeeds even though the client making the call, your HTTP client, isn't the same client the authorization issued the token to, the sample app.
```
Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
The call succeeds even though the client making the call (your HTTP client) isn't the same client the authorization issued the token to, the sample app.
```

Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
The call succeeds even though the client making the call, your HTTP client, isn't the same client the authorization issued the token to (the sample app).
```

File: _source/_posts/2024-09-10-angular-dpop-jwt.md [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+
+Let's call another endpoint with the same access token, the Okta Applications endpoint. Run the following HTTP request replacing `{yourOktaDomain}` and `{yourAccessToken}`:
+
+```http
+GET https://alisa.oktapreview.com/api/v1/apps HTTP/1.1
+Authorization: Bearer {yourAccessToken}
+```
+
+The call succeeds even though you call from a different client, like you saw in the prior step, calling the Users API. The call succeeds for a privileged user as long as the Okta Application has the `okta.apps.read` and the OIDC config has the scope. You may say that's a lot of constraints, and you're right. Okta adds a lot of guards when making API requests about the resources in the top-level Okta org, such as the list of Okta applications. This example demonstrates how powerful and vulnerable a privileged user, such as an admin, access is. Anyone with the token can make the same request, even if they are an attacker.
```
Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
The call succeeds even though you call from a different client, like you saw in the prior step, calling the Users API. The call succeeds for a privileged user as long as the Okta Application has the `okta.apps.read` and the OIDC config has the scope. You may say that's a lot of constraints, and you're right. Okta adds a lot of guards when making API requests about the resources in the top-level Okta org, such as the list of Okta applications. This example demonstrates how powerful and vulnerable a privileged user (such as an admin access) is. Anyone with the token can make the same request, even if they are an attacker.
```

File: _source/_posts/2024-09-10-angular-dpop-jwt.md [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+
+Back in the app, sign out to clear the authenticated session and tokens. We're making changes that require you to sign in from scratch.
+
+## Use secure coding techniques to protect your web apps
+
+All web applications must use secure coding techniques to protect from attacks, breaches, and malicious use. Public clients store their tokens within the user's hardware and require thoughtful security practices. Read more about SPA web security and security practices within Angular in this four-part series:
+
+{% excerpt /blog/2022/07/06/spa-web-security %}
+
+It doesn't matter if your application uses bearer tokens or DPoP; apps must employ secure coding practices. DPoP doesn't prevent attackers from stealing your token but constrains its use. DPoP uses asymmetric encryption to prove token ownership, so you must avoid exfiltration or unauthorized use of the keyset. If an attacker gets a hold of the private key, they can create valid proofs.
```
Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
It doesn't matter if your application uses bearer tokens or DPoP; apps must employ secure coding practices. DPoP doesn't prevent attackers from stealing your token but constrains its use. DPoP uses asymmetric encryption to prove token ownership, so you must avoid exfiltration or unauthorized use of the keyset. If an attacker gets a hold of the private key they can create valid proofs.
```

File: _source/_posts/2024-09-10-angular-dpop-jwt.md [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+```http
+HTTP/1.1 400 Bad Request
+
+{
+    "error": "invalid_dpop_proof",
+    "error_description": "The DPoP proof JWT header is missing."
+}
+```
+
+All HTTP requests to DPoP-protected resources, including the `/token` request to sign in, now require a DPoP proof. We must enable DPoP in the OIDC configuration.
```
Comment (coltonbeach) [inline _source/_posts/2024-09-10-angular-dpop-jwt.md]:
```suggestion
All HTTP requests to DPoP-protected resources (including the `/token` request to sign in) now require a DPoP proof. We must enable DPoP in the OIDC configuration.
```

**Review by coltonbeach — CHANGES_REQUESTED** [review]

File: _source/_posts/2024-09-10-angular-dpop-jwt.md (hunk 8 above)
Comment (alisaduncan) [reply _source/_posts/2024-09-10-angular-dpop-jwt.md]: I think I'll split this into 2 sentences

File: _source/_posts/2024-09-10-angular-dpop-jwt.md (hunk 11 above)
Comment (alisaduncan) [reply _source/_posts/2024-09-10-angular-dpop-jwt.md]: I'll reword this sentence entirely

File: _source/_posts/2024-09-10-angular-dpop-jwt.md (hunk 12 above)
Comment (alisaduncan) [reply _source/_posts/2024-09-10-angular-dpop-jwt.md]: Rewording

File: _source/_posts/2024-09-10-angular-dpop-jwt.md (hunk 14 above)
Comment (alisaduncan) [reply _source/_posts/2024-09-10-angular-dpop-jwt.md]: rewording

**Review by coltonbeach — APPROVED** [review]

**Pull request merged**